models/glif_psc.cpp

```diff
@@ -0,0 +1,578 @@
+      V_.t_ref_remaining_ -= dt;
+      if ( V_.t_ref_remaining_ <= 0.0 )
+      {
+        // Neuron has left refractory period, reset voltage and after-spike
+        // current
+
+        // Reset ASC_currentsfor glif3/4/5 models with "ASC"
+        if ( model_type == 3 || model_type == 4 || model_type == 5 )
+        {
+          for ( std::size_t a = 0; a < S_.ASCurrents_.size(); ++a )
+          {
+            S_.ASCurrents_[ a ] =
+              P_.asc_amps_[ a ] + S_.ASCurrents_[ a ] * P_.r_[ a ] * std::exp( -P_.k_[ a ] * V_.t_ref_total_ );
+          }
+        }
+
+        // Reset voltage
+        if ( model_type == 1 || model_type == 3 )
+        {
+          // Reset voltage for glif1/3 models without "R"
+          S_.U_ = P_.V_reset_;
+        }
+        else
+        {
+          // Reset voltage for glif2/4/5 models with "R"
+          S_.U_ = P_.voltage_reset_a_ * S_.U_ + P_.voltage_reset_b_;
+
+          // reset spike component of threshold
+          V_.last_spike_ = V_.last_spike_ + P_.a_spike_;
+
+          // rest the global threshold (voltage component of threshold: stay the
+          // same)
+          S_.threshold_ = V_.last_spike_ + V_.last_voltage_ + P_.th_inf_;
+        }
+        // Check if bad reset
+        // TODO: Better way to handle?
+        if ( S_.U_ > S_.threshold_ )
+        {
+          printf(
+            "Simulation Terminated: Voltage (%f) reset above threshold "
+            "(%f)!!\n",
+            S_.U_,
+            S_.threshold_ );
+        }
+        assert( S_.U_ <= S_.threshold_ );
+      }
+      else
+      {
+        S_.U_ = v_old;
+      }
+    }
+    else
+    {
+      // Integrate voltage and currents
+
+      // Calculate new ASCurrents value using expoential methods
+      S_.ASCurrents_sum_ = 0.0;
+      // for glif3/4/5 models with "ASC"
+      if ( model_type == 3 || model_type == 4 || model_type == 5 )
+      {
+        for ( std::size_t a = 0; a < S_.ASCurrents_.size(); ++a )
+        {
+          S_.ASCurrents_sum_ += S_.ASCurrents_[ a ];
+          S_.ASCurrents_[ a ] = S_.ASCurrents_[ a ] * std::exp( -P_.k_[ a ] * dt );
+        }
+      }
+      // voltage dynamics of membranes
+      // Linear Exact to find next V_m value
+      S_.U_ = v_old * V_.P33_ + ( S_.I_ + S_.ASCurrents_sum_ ) * V_.P30_;
+
+      // add synapse component for voltage dynamics
+      S_.I_syn_ = 0.0;
+      for ( size_t i = 0; i < P_.n_receptors_(); i++ )
+      {
+        S_.U_ += V_.P31_[ i ] * S_.y1_[ i ] + V_.P32_[ i ] * S_.y2_[ i ];
+        S_.I_syn_ += S_.y2_[ i ];
+      }
+
+      // Calculate exact voltage component of the threshold for glif5 model with
+      // "A"
+      if ( model_type == 5 )
+      {
+        double beta = ( S_.I_ + S_.ASCurrents_sum_ ) / P_.G_;
+        double phi = P_.a_voltage_ / ( P_.b_voltage_ - P_.G_ / P_.C_m_ );
+        voltage_component = phi * ( v_old - beta ) * std::exp( -P_.G_ * dt / P_.C_m_ )
```

Comment:
Why need for `voltage_compontent`? Why not update `S_.theta_voltage_` directly (see comment in h file for change in names).

Also, precompute all exponentials and divisions if at all possible. Can terms be rearranged for simpliciation and numerical stability?

Reply:
See comments on new state variables for spike and voltage components of threshold above. All exponential and divisions have been precomputed in <calibrate()>